portiaui/app/services/annotation-structure.js

```diff
@@ -270,24 +176,59 @@ const ElementStructure = Ember.Object.extend({
 
 const DataElementStructure = ElementStructure.extend({
     model: null,  // a sample
+    definition: [],
 
-    definition: Ember.computed('model.orderedAnnotations.[]', function() {
-        return (this.get('model.items') || []).filter(
-            item => !!item).map(function mapper(annotation) {
-                if (annotation.constructor.modelName === 'annotation') {
-                    return {
-                        annotation
-                    };
-                } else if (annotation.constructor.modelName === 'item') {
-                    return {
-                        annotation,
-                        children: (annotation.get('annotations') || []).map(mapper)
-                    };
-                }
-            });
-    })
+    setDefinition: Ember.on('init', Ember.observer('model.orderedAnnotations.[]', function() {
+        const sample = this.get('model');
+        if (!sample) {
+            this.set('definition', []);
+        }
+        const structurePromise = createStructure(sample);
+        this.currentPromise = structurePromise;
+        structurePromise.then(structure => {
+            if (structurePromise === this.currentPromise) {
+                delete this.currentPromise;
+                this.set('definition', structure);
+            }
+        });
+    }))
 });
 
+export function createStructure(sample) {
```

Comment:
this promise juggling is needed to get the mix of promises and PromiseManyArrays to resolve to a plain object at the end. if we split this up it will be even harder to see the structure that's expected as a result of this call.